components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/claims/impl/DefaultClaimHandler.java

```diff
@@ -1100,6 +1107,112 @@ private void handleRoleClaim(AuthenticationContext context, Map<String, String>
         }
     }
 
+    /**
+     * Add the application roles of federated user to remote claims.
+     *
+     * @param stepConfig       StepConfig of current step.
+     * @param context          AuthenticationContext of current authentication flow.
+     * @param idPClaimMappings Claim mappings of IdP of the current step.
+     * @param remoteClaims     Remote claims of the current step authenticated user.
+     * @throws FrameworkException Exception on handling application roles for federated user.
+     */
+    protected void handleApplicationRolesForFederatedUser(StepConfig stepConfig, AuthenticationContext context,
+                                                        ClaimMapping[] idPClaimMappings,
+                                                        Map<String, String> remoteClaims)
+            throws FrameworkException {
+
+        // IdP claim mappings should be available and the current step should be a subject attribute step.
+        if (idPClaimMappings == null || !stepConfig.isSubjectAttributeStep()) {
+            return;
+        }
+        // Get the remote claim URI of the application role claim.
+        String idpAppRoleClaimUri = Arrays.stream(idPClaimMappings)
+                .filter(claimMapping -> claimMapping.getLocalClaim().getClaimUri()
+                        .equals(FrameworkConstants.APP_ROLES_CLAIM))
+                .map(claimMapping -> claimMapping.getRemoteClaim().getClaimUri())
+                .findFirst()
+                .orElse(null);
+        // If there is no application role claim mapping, no need to proceed.
+        if (idpAppRoleClaimUri == null) {
+            return;
+        }
+        // Regardless of whether the application role claim is requested from the SP, we need to add it to the remote
+        // claims since otherwise we wouldn't know if application roles are resolved or not at a later stage.
+        AppRoleMappingConfig[] appRoleMappingConfigs = context.getSequenceConfig().getApplicationConfig()
+                .getServiceProvider().getApplicationRoleMappingConfig();
+        String fidpName = stepConfig.getAuthenticatedIdP();
+        boolean useAppRoleMapping = Arrays.stream(appRoleMappingConfigs)
+                .anyMatch(appRoleMappingConfig -> appRoleMappingConfig.getIdPName().equals(fidpName)
+                        && appRoleMappingConfig.isUseAppRoleMappings());
+        if (useAppRoleMapping) {
+            String appRoles = getApplicationRoles(stepConfig.getAuthenticatedUser(), context);
+            // Checking if the appRoles string is null but can be an empty string.
+            if (appRoles != null) {
```

Comment:
We can use StringUtils

Reply:
This can be an empty string if there are no roles available for the user.
The app role claim will still be added as an empty string.